### ISSUE TITLE
Lack of sending `flow` messages when remote windows are down to 0 cause panics.

### ISSUE THREAD

**Issue by rokeller** [minghuaw/fe2o3-amqp#170]
Currently, the session's handling of incoming and outgoing `transfer` messages always decrement the corresponding remote window size. At least for fe2o3-amqp receivers, this is an issue because once the `remote_outgoing_window` reached `0`, and the next `transfer` comes in, an integer underflow happens, causing the receiver to panic (see below). This results in a receiver being able to receive at most `outgoing-window` from the `begin`. It might be a good idea to send a `flow` performative at the latest when the window is down to `0`, or maybe even before then (e.g. given a threshold).

```
thread 'tokio-runtime-worker' panicked at 'attempt to subtract with overflow', /home/rogerk/code/rust/fe2o3-amqp/fe2o3-amqp/src/session/mod.rs:450:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
   2: core::panicking::panic
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:115:5
   3: <fe2o3_amqp::session::Session as fe2o3_amqp::endpoint::session::Session>::on_incoming_transfer::{{closure}}
             at /home/rogerk/code/rust/fe2o3-amqp/fe2o3-amqp/src/session/mod.rs:450:9
   4: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
   5: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/future.rs:124:9
   6: <fe2o3_amqp::acceptor::session::ListenerSession as fe2o3_amqp::endpoint::session::Session>::on_incoming_transfer::{{closure}}
             at /home/rogerk/code/rust/fe2o3-amqp/fe2o3-amqp/src/acceptor/session.rs:451:61
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
   8: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/future.rs:124:9
   9: fe2o3_amqp::session::engine::SessionEngine<S>::on_incoming::{{closure}}
             at /home/rogerk/code/rust/fe2o3-amqp/fe2o3-amqp/src/session/engine.rs:120:21
  10: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
  11: fe2o3_amqp::session::engine::SessionEngine<S>::event_loop::{{closure}}
             at /home/rogerk/code/rust/fe2o3-amqp/fe2o3-amqp/src/session/engine.rs:426:69
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
  13: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/future.rs:124:9
  14: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/core.rs:223:17
  15: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/loom/std/unsafe_cell.rs:14:9
  16: tokio::runtime::task::core::Core<T,S>::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/core.rs:212:13
  17: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:481:19
  18: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panic/unwind_safe.rs:271:9
  19: std::panicking::try::do_call
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:483:40
  20: __rust_try
  21: std::panicking::try
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:447:19
  22: std::panic::catch_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panic.rs:137:14
  23: tokio::runtime::task::harness::poll_future
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:469:18
  24: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:198:27
  25: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:152:15
  26: tokio::runtime::task::raw::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/raw.rs:255:5
  27: tokio::runtime::task::raw::RawTask::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/raw.rs:200:18
  28: tokio::runtime::task::LocalNotified<S>::run
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/mod.rs:459:9
  29: tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/worker.rs:487:21
  30: tokio::runtime::coop::with_budget
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/coop.rs:102:5
  31: tokio::runtime::coop::budget
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/coop.rs:68:5
  32: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/worker.rs:463:9
  33: tokio::runtime::scheduler::multi_thread::worker::Context::run
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/worker.rs:426:24
  34: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/worker.rs:406:17
  35: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/macros/scoped_tls.rs:61:9
  36: tokio::runtime::scheduler::multi_thread::worker::run
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/worker.rs:403:5
  37: tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/worker.rs:365:45
  38: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/blocking/task.rs:42:21
  39: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/core.rs:223:17
  40: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/loom/std/unsafe_cell.rs:14:9
  41: tokio::runtime::task::core::Core<T,S>::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/core.rs:212:13
  42: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:481:19
  43: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panic/unwind_safe.rs:271:9
  44: std::panicking::try::do_call
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:483:40
  45: __rust_try
  46: std::panicking::try
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:447:19
  47: std::panic::catch_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panic.rs:137:14
  48: tokio::runtime::task::harness::poll_future
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:469:18
  49: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:198:27
  50: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/harness.rs:152:15
  51: tokio::runtime::task::raw::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/raw.rs:255:5
  52: tokio::runtime::task::raw::RawTask::poll
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/raw.rs:200:18
  53: tokio::runtime::task::UnownedTask<S>::run
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/task/mod.rs:496:9
  54: tokio::runtime::blocking::pool::Task::run
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/blocking/pool.rs:159:9
  55: tokio::runtime::blocking::pool::Inner::run
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/blocking/pool.rs:510:17
  56: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/rogerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/blocking/pool.rs:468:13
```

**Comment by minghuaw** [comment]
Thanks for raising the issue. I will look at it today

**Comment by minghuaw** [comment]
I took a quick look at the spec and found what I forgot to implement in the session

> The remote-outgoing-window reflects the maximum number of incoming transfers that MAY arrive without exceeding the remote endpoint’s outgoing-window. This value MUST be decremented after every incoming transfer frame is received, and recomputed when in- formed of the remote session endpoint state. When this window shrinks, it is an indication of outstanding transfers. Settling outstanding transfers can cause the window to grow.

I forgot to increment this value when settling messages, and of course I should have used a `checked_sub` instead

**Comment by rokeller** [comment]
As far as I can tell, the `window-violation` error (see [Session Errors](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-session-error)) is meant to be sent in case the session's senders would send too many `transfer` performatives, even when accounting for settled transfers.

**Comment by minghuaw** [comment]
I think that is meant for violation of local incoming window, which is actually not checked right now

**Comment by minghuaw** [comment]
> I think that is meant for violation of local incoming window, which is actually not checked right now

I will probably add code checking for incoming window

**Comment by minghuaw** [comment]
I think why I didn't implement checking the local incoming window was because the spec says

> and MAY (depending on policy) decrement its incoming-window.

I will probably leave this for some future release instead of the next one which will solely focus on avoiding the overflow/underflow

**Comment by minghuaw** [comment]
@rokeller I have just done a quick review over all `+=` and `-=` in branch `#170-overflow`. Searching `+` or `-` will return a ton of irrelevant result and making reviewing them almost impossible. Would you mind giving this branch a try before I publishing a new release? You might need to set all `fe2o3-amqp` deps to git, the toml file would look like

```toml
fe2o3-amqp = { git = "https://github.com/minghuaw/fe2o3-amqp", branch = "#170-overflow" }

# if you need other deps
fe2o3-amqp-management = { git = "https://github.com/minghuaw/fe2o3-amqp", branch = "#170-overflow" }
```

**Comment by rokeller** [comment]
Hi,

I've run some tests sending > 15K messages and receiving > 30K messages and didn't get any panics anymore. I can probably run more tests tomorrow, if needed.

**Comment by minghuaw** [comment]
@rokeller Thanks for confirming. I will make it a new release soon then.

**Comment by minghuaw** [comment]
Hi @rokeller 

The fix has been applied in the new release (0.7.19 and 0.8.4). I will close this for now. Feel free to re-open the issue should there be any problem or questions
